pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -281,7 +281,15 @@ func syncVolume(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCl
 		if volume.Spec.ClaimRef == nil {
 			return fmt.Errorf("PersistentVolume[%s] expected to be bound but found nil claimRef: %+v", volume.Name, volume)
 		} else {
-			_, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+			claim, err := binderClient.GetPersistentVolumeClaim(volume.Spec.ClaimRef.Namespace, volume.Spec.ClaimRef.Name)
+
+			// a claim can be deleted and recreated with the same name before the controller's sync period
+			// has a chance to release the volume.  Finding the claim by name/namespace above might not actually
+			// get the original object, so a UID check is necessary.
+			if claim != nil && claim.UID != volume.Spec.ClaimRef.UID {
+				nextPhase = api.VolumeReleased
+			}
+
```

Comment:
I'd slightly prefer error check before the claim check, i.e.:
```
if err != nil {
...
} else if claim.UID != ... {
}
```

It's just matter of style, the code is readable even now.